python/examples/panorama_sfm.py

```diff
@@ -0,0 +1,222 @@
+                raise ValueError(
+                    "Panoramas of different sizes are not supported."
+                )
+
+        for cam_idx, cam_from_pano_r in enumerate(cams_from_pano_rotation):
+            rays_in_pano = rays_in_cam @ cam_from_pano_r
+            xy_in_pano = spherical_img_from_cam(pano_size, rays_in_pano)
+            xy_in_pano = xy_in_pano.reshape(
+                camera.width, camera.height, 2
+            ).astype(np.float32)
+            xy_in_pano -= 0.5  # COLMAP to OpenCV pixel origin.
+            image = cv2.remap(
+                pano_image,
+                *np.moveaxis(xy_in_pano, -1, 0),
+                cv2.INTER_LINEAR,
+                borderMode=cv2.BORDER_WRAP,
+            )
+            # We define a mask such that each pixel of the panorama has its
+            # features extracted only in a single virtual camera.
+            closest_camera = np.argmax(rays_in_pano @ cam_centers_in_pano.T, -1)
+            mask = (
+                ((closest_camera == cam_idx) * 255)
+                .astype(np.uint8)
+                .reshape(camera.width, camera.height)
+            )
+
+            image_name = rig_config.cameras[cam_idx].image_prefix + pano_name
+            mask_name = f"{image_name}.png"
+
+            image_path = output_image_dir / image_name
+            image_path.parent.mkdir(exist_ok=True, parents=True)
+            if not pycolmap.Bitmap.from_array(image).write(image_path):
+                raise RuntimeError(f"Cannot write {image_path}")
+
+            mask_path = mask_dir / mask_name
+            mask_path.parent.mkdir(exist_ok=True, parents=True)
+            if not pycolmap.Bitmap.from_array(mask).write(mask_path):
+                raise RuntimeError(f"Cannot write {mask_path}")
+
+    return rig_config
+
+
+def run(args):
+    # Define the paths.
+    image_dir = args.output_path / "images"
+    mask_dir = args.output_path / "masks"
+    image_dir.mkdir(exist_ok=True, parents=True)
+    mask_dir.mkdir(exist_ok=True, parents=True)
+
+    database_path = args.output_path / "database.db"
+    if database_path.exists():
+        database_path.unlink()
+
+    rec_path = args.output_path / "sparse"
+    rec_path.mkdir(exist_ok=True, parents=True)
+
+    # Search for input images.
+    pano_image_dir = args.input_image_path
+    pano_image_names = sorted(
+        p.relative_to(pano_image_dir).as_posix()
+        for p in pano_image_dir.rglob("*")
+        if not p.is_dir()
+    )
+    logging.info(f"Found {len(pano_image_names)} images in {pano_image_dir}.")
+
+    rig_config = render_perspective_images(
+        pano_image_names, pano_image_dir, image_dir, mask_dir
+    )
+    pycolmap.set_random_seed(0)
+    pycolmap.extract_features(
+        database_path,
+        image_dir,
+        reader_options={"mask_path": mask_dir},
+        camera_mode=pycolmap.CameraMode.PER_FOLDER,
+    )
+
+    db = pycolmap.Database(database_path)
```

Comment:
nit: We can now use the database with a context manager as:
```
with pycolmap.Database(database_path) as db:
    ...
```

Reply:
I forgot about this, thank you.